probe/endpoint/nat.go

```diff
@@ -0,0 +1,129 @@
+package endpoint
+
+import (
+	"encoding/xml"
+	"io"
+	"log"
+	"os/exec"
+	"strconv"
+
+	"github.com/weaveworks/scope/report"
+)
+
+// these structs are for the parsed conntrack output
+type layer3 struct {
+	XMLName xml.Name `xml:"layer3"`
+	SrcIP   string   `xml:"src"`
+	DstIP   string   `xml:"dst"`
+}
+
+type layer4 struct {
+	XMLName xml.Name `xml:"layer4"`
+	SrcPort int      `xml:"sport"`
+	DstPort int      `xml:"dport"`
+	Proto   string   `xml:"protoname,attr"`
+}
+
+type meta struct {
+	XMLName   xml.Name `xml:"meta"`
+	Direction string   `xml:"direction,attr"`
+	Layer3    layer3   `xml:"layer3"`
+	Layer4    layer4   `xml:"layer4"`
+}
+
+type flow struct {
+	XMLName xml.Name `xml:"flow"`
+	Metas   []meta   `xml:"meta"`
+}
+
+type conntrack struct {
+	XMLName xml.Name `xml:"conntrack"`
+	Flows   []flow   `xml:"flow"`
+}
+
+// This is our 'abstraction' of the conntract output
+// original is the private IP that has been rewritten
+type connection struct {
+	originalIP   string
+	originalPort int
+
+	rewrittenIP   string
+	rewrittenPort int
+}
+
+func natTable() ([]connection, error) {
+	var conntrack conntrack
+	cmd := exec.Command("conntrack", "-L", "--any-nat", "-o", "xml")
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		return nil, err
+	}
+	if err := cmd.Start(); err != nil {
+		return nil, err
+	}
+	if err := xml.NewDecoder(stdout).Decode(&conntrack); err != nil {
+		if err == io.EOF {
+			return []connection{}, nil
+		}
+		return nil, err
+	}
+	if err := cmd.Wait(); err != nil {
+		return nil, err
+	}
+
+	output := []connection{}
+	for _, flow := range conntrack.Flows {
+		original, reply := meta{}, meta{}
+		for _, meta := range flow.Metas {
+			if meta.Direction == "original" {
+				original = meta
+			} else if meta.Direction == "reply" {
+				reply = meta
+			}
+		}
+
+		if original.Layer4.Proto != "tcp" {
+			continue
+		}
+
+		var conn connection
+		if original.Layer3.SrcIP == reply.Layer3.DstIP {
+			conn = connection{
+				originalIP:    reply.Layer3.SrcIP,
+				originalPort:  reply.Layer4.SrcPort,
+				rewrittenIP:   original.Layer3.DstIP,
+				rewrittenPort: original.Layer4.DstPort,
+			}
+		} else {
+			conn = connection{
+				originalIP:    original.Layer3.SrcIP,
+				originalPort:  original.Layer4.SrcPort,
+				rewrittenIP:   reply.Layer3.DstIP,
+				rewrittenPort: reply.Layer4.DstPort,
+			}
+		}
+
+		output = append(output, conn)
+	}
+
+	return output, nil
+}
+
+func applyNAT(scope string, rpt report.Report) {
```

Comment:
I'd prefer report as the first argument (since it gets modified), but not critical.